sycl/plugins/unified_runtime/ur/adapters/cuda/event.cpp

```diff
@@ -0,0 +1,309 @@
+//===--------- event.cpp - CUDA Adapter ------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===-----------------------------------------------------------------===//
+
+#include "event.hpp"
+#include "common.hpp"
+#include "context.hpp"
+#include "device.hpp"
+#include "queue.hpp"
+
+#include <cassert>
+#include <cuda.h>
+
+ur_event_handle_t_::ur_event_handle_t_(ur_command_t type,
+                                       ur_context_handle_t context,
+                                       ur_queue_handle_t queue, CUstream stream,
+                                       uint32_t stream_token)
+    : commandType_{type}, refCount_{1}, has_ownership_{true},
+      hasBeenWaitedOn_{false}, isRecorded_{false}, isStarted_{false},
+      streamToken_{stream_token}, evEnd_{nullptr}, evStart_{nullptr},
+      evQueued_{nullptr}, queue_{queue}, stream_{stream}, context_{context} {
+
+  bool profilingEnabled = queue_->ur_flags_ & UR_QUEUE_FLAG_PROFILING_ENABLE;
+
+  UR_CHECK_ERROR(cuEventCreate(
+      &evEnd_, profilingEnabled ? CU_EVENT_DEFAULT : CU_EVENT_DISABLE_TIMING));
+
+  if (profilingEnabled) {
+    UR_CHECK_ERROR(cuEventCreate(&evQueued_, CU_EVENT_DEFAULT));
+    UR_CHECK_ERROR(cuEventCreate(&evStart_, CU_EVENT_DEFAULT));
+  }
+
+  if (queue_ != nullptr) {
+    urQueueRetain(queue_);
+  }
+  urContextRetain(context_);
+}
+
+ur_event_handle_t_::ur_event_handle_t_(ur_context_handle_t context,
+                                       CUevent eventNative)
+    // TODO(ur): Missing user command type
```

Comment:
Mostly a question than something to action:
Assuming that for the CUDA backend, `UR_COMMAND_EVENTS_WAIT` command will do be okay, but is it planned to have createUserEvent type of command in UR? 
Also, I was thinking this may show up later when dealing with the OpenCL backend port, despite just being an implementation detail eod.

Reply:
We've deliberately removed support for user events from UR (https://github.com/oneapi-src/unified-runtime/pull/100) since they are not used and are also a pain to implement.

Does that may make this todo redundant?

Reply:
Oh I see, I just followed up the discussion regarding this. If the intent is to completely rid of this in the future anyways, as it seems may not even fit all platforms efficiently, and the current alternative way of enqueueing a wait command sync looks fine. If there is not a future plan of reworking that , then maybe the _TODO_ is just reduntant.
 @JackAKirk @npmiller Wonder what you think.

Reply:
> Sounds like the meaningful question is "Is a cross platform event wait useful/required?"?

That's not even possible with [clCreateUserEvent](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#clCreateUserEvent) which can only synchronize within a context within a platform and not across platforms.

My view on this topic is, if/when that question gets raised by users, is to have an in depth investigation about how to actually support this type of functionality in a way that works for all adapters because they will need to share some kind of synchronization contract which currently does not exist.

Reply:
In that case I think it definitely makes sense to remove it, since sycl::context only has any meaning for opencl (and even then it seems it is mostly unused). In cuda and l0 backends users will just use the default sycl context which will have all devices in each platform in that default context. The supposed purpose of sycl context, to control memory accesses between different devices within a sycl::context, is not practically enforceable for backends like cuda; instead this is the role of extensions like the USM P2P extension, which can support the management of constraints such as a maximum number of peer devices.